### ISSUE TITLE
Printer DCP-L2560DW works PARTIALLY with driver "Brother DCP-L2540DW series, using brlaser v6"

### ISSUE THREAD

**Issue by jmickelin** [pdewacht/brlaser#210]
I simply could not get the `lpinfo` command to actually show any relevant output, even though the printer once configured was clearly visible with `lpstat`. I think it's because of my network setup (printer being on a different VLAN) and some possible firewall interference in between.

Instead, I used `/usr/lib/cups/backend/snmp $printers_ip_address` as described in [this old Fedora bug thread](https://bugzilla.redhat.com/show_bug.cgi?id=583909#c9). I hope that's alright!

This is the output I got:
```
network lpd://skrifvaren330A5/BINARY_P1 "Brother DCP-L2560DW series" "Brother DCP-L2560DW series" "MFG:Brother;CMD:PJL,PCL,PCLXL,URF;MDL:DCP-L2560DW series;CLS:PRINTER;CID:Brother Laser Type1;URF:W8,CP1,IS4-1,MT1-3-4-5-8,OB10,PQ4,RS300-600,V1.3,DM1;" "Funkytown"
```

I think this would correspond to the following fields (had I got `lpinfo` to work):
```
Device: uri = lpd://skrifvaren330A5/BINARY_P1
class = network
info = Brother DCP-L2560DW series
make-and-model = Brother DCP-L2560DW series
device-id = MFG:Brother;CMD:PJL,PCL,PCLXL,URF;MDL:DCP-L2560DW series;CLS:PRINTER;CID:Brother Laser Type1;URF:W8,CP1,IS4-1,MT1-3-4-5-8,OB10,PQ4,RS300-600,V1.3,DM1;
location = Funkytown
```

Not sure which one of the third and fourth fields correspond to `info` and `make-and-model`, but luckily they are identical for this model!

**Comment by jmickelin** [comment]
**Correction**:

It works fine at 600 DPI (tested in both portrait and landscape, simplex and duplex), but does not work at 1200HQ (i.e. 2400x600 according to the [product specification](https://support.brother.com/g/b/spec.aspx?c=se&lang=sv&prod=dcpl2560dw_eu)). It does not properly shrink the content onto the page, and instead prints it large and cropped.

Furthermore, the "scale" setting is grayed out and cannot be changed in the GUI.

**Comment by QORTEC** [comment]
Thank you for your report, your printer has been added to my pull request.

**Comment by jmickelin** [comment]
Thanks! Fingers crossed it gets merged at some point in the future :)

I noticed I had mistyped the model name in the title of this issue (but not in the command output). Indeed, it is a DCP-L2560DW, like what you added in your PR. I'll edit the title here, if nothing else, just to help others googling it in the future.

**Comment by QORTEC** [comment]
I'm maintaining a fork of `brlaser` at `Owl-Maintain`, which packagers are starting to use.

So far I know of; FreeBSD, Arch, Gentoo, Solus, and I've heard even Debian is switching.

So if you run into any bugs you may want to report it there.

**Comment by QORTEC** [comment]
There is new code in `Owl-Maintain/brlaser` that may resolve your issue printing at 1200dpi, though v2.6.8 was just released so it may take some time before it arrives by your package manager.

**Comment by jmickelin** [comment]
@QORTEC Hi! Sorry for the late reply. I finally had the time to check out your fork. I'm running v2.6.8, but unfortunately both the issues I saw before are still there.

The "scale" option is greyed out in system's print dialog:
<img width="604" height="483" alt="Image" src="https://github.com/user-attachments/assets/68e7047b-3bb9-4bcb-bda5-f768bcdc2b96" />

---

The resolution settings show only 600dpi and 1200x600dpi. Though the GUI calls it HQ1200, for some reason. I thought that was the name of the 2400x600dpi one, at least according to your [definitions for the Test drivers](https://github.com/Owl-Maintain/brlaser/blob/864410484a6c8486f52573144b3337ca1dd75fed/brlaser.drv.in#L111).

<img width="451" height="174" alt="Image" src="https://github.com/user-attachments/assets/88d8b5a7-70da-4722-b0e0-ab9e09779b97" />

(Though it seems that [another part of that file](https://github.com/Owl-Maintain/brlaser/blob/864410484a6c8486f52573144b3337ca1dd75fed/brlaser.drv.in#L48) does use HQ1200 to refer to 1200x600dpi. Is this a mistake?)

And this printer should be using 2400x600dpi, according to the manufacturer's product page:

<img width="774" height="148" alt="Image" src="https://github.com/user-attachments/assets/e5972192-3ac3-4532-ac44-a254aa15e84e" />

---

Finally, when printing out a test page, both this and the proprietary driver consistently ends up printing ever so slightly off-center, compared to the IPP everywhere/driverless driver:
![Image](https://github.com/user-attachments/assets/90927e19-41df-4bee-9a4e-8472605ca582)
![Image](https://github.com/user-attachments/assets/6766742e-57b6-442e-9204-48e9f65779f5)
![Image](https://github.com/user-attachments/assets/7844b94e-34ed-41fc-a492-b4bd95cdb966)

**Comment by jmickelin** [comment]
Actually, if I tell it to use the `Brother A Owl-Maintain/brlaser Test Driver` and select the `HQ1200 (2400x600dpi)` resolution, it does print out at the correct scale. Still offset like above, but this time to the right.

Also, notice how all the three drivers get different Media Limits? The proprietary one differs only slightly from driverless, while the brlaser one is way off. Not sure if this means anything, but I wanted to point it out.

**Comment by jmickelin** [comment]
Sorry to triple post, but I found out that you can request the settings that the printer self-reports for IPP Everywhere and dump them to a file with this command:
```
$ ipptool -tv ipp://printer-address/ipp/print get-printer-attributes.test >attributes.txt
```

Here is what my printer gives me: [attributes.txt](https://github.com/user-attachments/files/24061417/attributes.txt)

One thing I noticed is that it reports the margins for the A4 size to be `0.432cm`, which matches up pretty well with the Media Limits of the IPP Everywhere test page (for the maximum values, compare them to the dimensions of an A4 paper: 21.0cm - 0.432cm = 20.568cm ≈ 20.57cm; likewise 29.7cm - 0.432cm = 29.268cm ≈ 29.27cm).

So it looks like IPP Everywhere correctly reads the reported margins, but not the drivers. Let's look at those:

The proprietary driver sets the `HWMargins` (and likewise the lower margins in `ImageableArea A4/A4`) to 12pt (4.2333mm), and the upper ones in `ImageableArea` are also off by the same factor. Could this be the cause of the offset? It's shifted by more than a tenth of a millimeter, for sure. But I noticed that the frame on the proprietary driver's test page is a bit wider than that of the IPP Everywhere one, so my theory is that it affects scaling too?

As for the `brlaser` driver, the margins are set to `8 8 8 16` pts, which is wildly off. So maybe it's as simple as changing these? [According to the documentation](https://www.cups.org/doc/ref-ppdcfile.html#HWMargins), the values don't have to be given in pts either, so you could use the exact measurements in centimeters, without any rounding:
```
HWMargins 0.432cm 0.432cm 0.432cm 0.432cm
```

Of course, it's not guaranteed that this will be the same size supported by all printers, but with this new `ipptool` dump, you could probably get a lot more interesting information from other users to set the correct value on a per-printer basis! I'm not sure whether all the old models support IPP Everywhere, but many will.

---

One more thing I noticed in my dump is that it does list 300dpi as being supported, so that could also be added for this printer in your driver.

**Comment by QORTEC** [comment]
@jmickelin Thank you for investigating this issue, the ipp attributes command you pointed out will definitely be helpful. 

I will definitely look into how `brlaser` handles margins and see what can be done.

I also still need to lookup each printer and check which 1200dpi mode it supports, then enable it in brlaser.

**Comment by jmickelin** [comment]
Makes sense! If the other product pages are as detailed as the one for my printer, maybe it's possible to scrape them for the info about what 1200dpi mode is used. Though there is a problem with not all pages being available in all languages (hence why I linked and screenshot the Swedish version; there wasn't any English-language page when I looked for it)...
